core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -2515,4 +2510,214 @@ public RexNode simplifyFilterPredicates(Iterable<? extends RexNode> predicates)
+      switch (kind) {
+      case LESS_THAN:
+        b.addRange(Range.lessThan(value), literal.getType());
+        return true;
+      case LESS_THAN_OR_EQUAL:
+        b.addRange(Range.atMost(value), literal.getType());
+        return true;
+      case GREATER_THAN:
+        b.addRange(Range.greaterThan(value), literal.getType());
+        return true;
+      case GREATER_THAN_OR_EQUAL:
+        b.addRange(Range.atLeast(value), literal.getType());
+        return true;
+      case EQUALS:
+        b.addRange(Range.singleton(value), literal.getType());
+        return true;
+      case NOT_EQUALS:
+        b.addRange(Range.lessThan(value), literal.getType());
+        b.addRange(Range.greaterThan(value), literal.getType());
+        return true;
+      case SEARCH:
+        final Sarg sarg = literal.getValueAs(Sarg.class);
+        b.addSarg(sarg, negate, literal.getType());
+        return true;
+      case IS_NULL:
+        if (negate) {
+          throw new AssertionError();
+        }
+        b.containsNull = true;
+        return true;
+      default:
+        throw new AssertionError("unexpected " + kind);
+      }
+    }
+
+    /** If a term is a call to {@code SEARCH} on a {@link RexSargBuilder},
+     * converts it to a {@code SEARCH} on a {@link Sarg}. */
+    RexNode fix(RexBuilder rexBuilder, RexNode term) {
+      if (term instanceof RexSargBuilder) {
+        RexSargBuilder sargBuilder = (RexSargBuilder) term;
+        return rexBuilder.makeCall(SqlStdOperatorTable.SEARCH, sargBuilder.ref,
+            rexBuilder.makeSearchArgumentLiteral(sargBuilder.build(negate),
+                term.getType()));
+      }
+      return term;
+    }
+  }
+
+  /** Equivalent to a {@link RexLiteral} whose value is a {@link Sarg},
+   * but mutable, so that the Sarg can be expanded as {@link SargCollector}
+   * traverses a list of OR or AND terms.
+   *
+   * <p>The {@link SargCollector#fix} method converts it to an immutable
+   * literal. */
+  static class RexSargBuilder extends RexNode {
+    final RexNode ref;
+    private final RexBuilder rexBuilder;
+    private List<RelDataType> types = new ArrayList<>();
+    final RangeSet<Comparable> rangeSet = TreeRangeSet.create();
+    boolean containsNull;
+
+    RexSargBuilder(RexNode ref, RexBuilder rexBuilder, boolean negate) {
+      this.ref = Objects.requireNonNull(ref);
+      this.rexBuilder = Objects.requireNonNull(rexBuilder);
+      this.containsNull = false;
+    }
+
+    @Override public String toString() {
+      return "SEARCH(" + ref + ", " + rangeSet
+          + (containsNull ? " + null)" : ")");
+    }
+
+    /** Returns a rough estimate of whether it is worth converting to a Sarg.
+     *
+     * <p>Examples:
+     * <ul>
+     *   <li>{@code x = 1}, {@code x <> 1}, {@code x > 1} have complexity 1
+     *   <li>{@code x > 1 or x is null} has complexity 2
+     *   <li>{@code x in (2, 4, 6) or x > 20} has complexity 4
```

Comment:
Should we take into account the cost of the logical operations?

Reply:
I think almost any weighting is good enough here, and the proper one should account for the data distribution (and the indices available), so we should not try making this costing too sophisticated.

Reply:
This is complexity, not cost. Logical operations don't have a cost! Physical operations would assign a different cost to, say, `x = 10 or x between 999 and 1001,` depending on whether `x` was stored in a hash map or sorted list.

I just wanted to ensure that we didn't use a `Sarg` for really simple comparisons, so I agree with @vlsi.